### PR TITLE
Basic Kobo Support

### PR DESCRIPTION
**Pull request by ragusa87**
* Allow syncing some specifics shelves to your Kobo
* So far, there is some DB changes and migration stuff to be discussed.
* All secret reported by bitguardian where false-positive from examples json. I could not ignore them (https://github.com/GitGuardian/ggshield/issues/548)  so I rebased the MR without theses files.
* I had trouble with using migration as the first "schema creation" was not into a migration, I fixed it.

### CODE REVIEW

File: config/packages/security.yaml [hunk 1]
```diff
@@ -37,12 +46,14 @@ security:
...
             # switch_user: true
 
+
     # Easy way to control access for large sections of your site
     # Note: Only the *first* access control that matches will be used
     access_control:
         - { path: ^/login, roles: PUBLIC_ACCESS }
         - { path: ^/logout, roles: ROLE_USER }
         - { path: ^/, roles: ROLE_USER }
+        - { path: ^/kobo, roles: PUBLIC_ACCESS }
```
Comment (SergioMendolia) [inline config/packages/security.yaml:56]: J'aime pas du tout l'idée d'avoir cette api ouverte au monde

File: rector.php
Comment (SergioMendolia) [inline rector.php:1]: 🤨

File: src/Controller/KoboNextReadController.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    {
+    }
+
+    #[Route('/v1/products/{uuid}/nextread', requirements: ['uuid' => '^[a-zA-Z0-9\-]+$'], methods: ['GET', 'POST'])]
+    public function nextRead(Request $request): Response
+    {
+        if ($this->koboStoreProxy->isEnabled()) {
+            return $this->koboStoreProxy->proxyOrRedirect($request);
+        }
+
```
Comment (SergioMendolia) [inline src/Controller/KoboNextReadController.php:25]: nice feature to implement

File: src/Entity/Kobo.php
Comment (SergioMendolia) [inline src/Entity/Kobo.php:1]: mv KoboDevice ?

File: src/Kobo/DownloadHelper.php [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    public function getCoverSize(Book $book): int
+    {
+        return $this->fileSystemManager->getCoverSize($book) ?? 0;
+    }
+
+    public function isEpub3(Book $book): bool
+    {
+        return $book->getExtension() === 'epub3' || $this->readEpubVersionIs3($book) === true;
```
Comment (SergioMendolia) [inline src/Kobo/DownloadHelper.php:42]: ?

File: src/Kobo/DownloadHelper.php [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        $zip = new \ZipArchive();
+
+        if ($zip->open($this->getBookFilename($book)) !== true) {
+            $this->logger->debug('Unable to open epub file to detect the format', ['book' => $book->getId()]);
+
+            return null;
+        }
+        try {
+            // Check for EPUB version
```
Comment (SergioMendolia) [inline src/Kobo/DownloadHelper.php:131]: php ebook should manage that

File: src/Kobo/ImageProcessor/CoverTransformer.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            $this->fallback($coverPath);
+
+            return;
+        }
+
+        // We only support jpeg & png & gif
+        if (false === str_ends_with($coverPath, '.jpg')
+            && false === str_ends_with($coverPath, '.png')
+            && false === str_ends_with($coverPath, '.gif')
+            && false === str_ends_with($coverPath, '.jpeg')
```
Comment (SergioMendolia) [inline src/Kobo/ImageProcessor/CoverTransformer.php:29]: constant

File: src/Kobo/ImageProcessor/CoverTransformer.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+        // Resize image based on the width and height
+        $sizes = getimagesize($coverPath);
+        if ($sizes === false) {
+            $this->fallback($coverPath);
+
+            return;
+        }
+        list($originalWidth, $originalHeight) = $sizes;
+
```
Comment (SergioMendolia) [inline src/Kobo/ImageProcessor/CoverTransformer.php:44]: why not use imaginebundle?

File: src/Menu/MenuBuilder.php [hunk 7]
```diff
@@ -77,6 +77,7 @@ public function createMainMenu(array $options): ItemInterface
         if ($this->security->isGranted('ROLE_ADMIN')) {
             $menu->addChild('admin_divider', ['label' => 'Admin'])->setExtra('divider', true);
             $menu->addChild('Admin', ['route' => 'app_user_index', ...$this->defaultAttr])->setExtra('icon', 'gear-fill');
+            $menu->addChild('Kobos', ['route' => 'app_kobo_admin_index', ...$this->defaultAttr])->setExtra('icon', 'gear-fill');
```
Comment (SergioMendolia) [inline src/Menu/MenuBuilder.php:80]: move to be user param

**Review by SergioMendolia — CHANGES_REQUESTED** [review]
A discuter ensemble

File: src/Entity/Kobo.php
Comment (ragusa87) [reply src/Entity/Kobo.php:1]: Done

File: src/Kobo/DownloadHelper.php (hunk 3 above)
Comment (ragusa87) [reply src/Kobo/DownloadHelper.php:42]: Indeed this is not used anymore, I will remove it

Comment (ragusa87) [reply src/Kobo/DownloadHelper.php]: done.

File: src/Kobo/ImageProcessor/CoverTransformer.php (hunk 6 above)
Comment (ragusa87) [reply src/Kobo/ImageProcessor/CoverTransformer.php:51]: We can discuss this, but it seems complicated to me.

File: config/packages/security.yaml (hunk 1 above)
Comment (ragusa87) [reply config/packages/security.yaml:56]: A discuter, dans tous les cas on check que le paramètre /kobo/{accessKey} soit existant pour identifier l'utilisateur.

C'est sûrement possible de faire un "PreAuthenticatedToken" avec un rôle dédié, faudrait creuser ça.

File: src/Kobo/DownloadHelper.php (hunk 4 above)
Comment (ragusa87) [reply src/Kobo/DownloadHelper.php]: I removed this part anyway..

File: src/Menu/MenuBuilder.php (hunk 7 above)
Comment (ragusa87) [reply src/Menu/MenuBuilder.php]: Done

File: config/packages/security.yaml (hunk 1 above)
Comment (ragusa87) [reply config/packages/security.yaml]: J'avais déjà fait tout ça en fait. Du coup, j'ai fixé.

**Comment by SergioMendolia** [comment]
closed because merge in another pr

**Pull request closed without merging**